python/ray/data/read_api.py

```diff
@@ -259,17 +259,16 @@ def remote_read(i: int, task: ReadTask, stats_actor) -> MaybeBlockPartition:
 
     if ray_remote_args is None:
         ray_remote_args = {}
-    # Increase the read parallelism by default to maximize IO throughput. This
-    # is particularly important when reading from e.g., remote storage.
-    if "num_cpus" not in ray_remote_args:
-        # Note that the too many workers warning triggers at 4x subscription,
-        # so we go at 0.5 to avoid the warning message.
-        ray_remote_args["num_cpus"] = 0.5
```

Comment:
Since we found this 0.5 CPU resource request to be much more ideal for these IO-bound read tasks, it would be nice if we could preserve this resource request when no stage fusion is taking place.

What if we merged resource requests when fusing stages, taking the maximum when encountering conflicts?
```python
def fuse(self, prev: Stage):
    # ...
    ray_remote_args = self._merge_ray_remote_args(prev.ray_remote_args)
    return OneToOneStage(name, block_fn, prev.compute, ray_remote_args)

def _merge_ray_remote_args(ray_remote_args: Dict[str, Any]) -> Dict[str, Any]:
    if "num_cpus" in ray_remote_args or "num_cpus" in self.ray_remote_args:
        ray_remote_args["num_cpus"] = max(
            ray_remote_args.get("num_cpus", default_num_cpus),
            self.ray_remote_args.get("num_cpus", default_num_cpus))
    # Same for num_gpus, memory, etc.
    self_resources = self.ray_remote_args.get("resources")
    if self_resources is not None:
        if "resources" not in ray_remote_args:
            ray_remote_args["resources"] = {}
        resources = ray_remote_args["resources"]
        for key, resource in self_resources:
            resources][key] = max(resource, resources[key]) if key in resources else resource
    return ray_remote_args
```

Reply:
Hmm, one issue is that this will make it impossible to actually specify 0.5 explicitly in a pipeline (which currently disables fusion).

Maybe we can revisit this later depending on use cases.